h/services/group_update.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+from sqlalchemy.exc import SQLAlchemyError, IntegrityError
+
+from h.exceptions import APIError, ConflictError
+
+
+class GroupUpdateService(object):
+
+    def __init__(self, session):
+        """
+        Create a new GroupUpdateService
+
+        :param session: the SQLAlchemy session object
+        """
+        self.session = session
+
+    def update(self, group, data):
+        """
+        Update a group model with the data provided.
+
+        This assumes that ``data`` has already been validated through a schema
+        or other methods. This method will intercept known SQL/Integrity errors
+        but the caller should be cautious not to expose unhandled Exceptions.
+
+        :raises ConflictError: If a provided ``authority_provided_id`` violates a
+                               DB uniqueness constraint. See :py:`h.views.api.exceptions`
+                               for a handler view for this error
+        :raises APIError:      If an unexpected SQLAlchemy error is encountered, to
+                               avoid exposing an SQL-specific Exception to users/clients
+        :raises ValueError:    (via model) if attributes fail model-level validation constraints
+        """
+
+        for key, value in data.items():
+            setattr(group, key, value)
+
+        try:
+            self.session.flush()
+
+        except IntegrityError as err:
+            if "group__groupid" in repr(err):
+                raise ConflictError(
+                    "authority_provided_id '{id}' is already in use".format(id=data['authority_provided_id']))
+
+        except SQLAlchemyError:
+            raise APIError('The update was not successful')
```

Comment:
I have no idea what this message should say, really, but the point being we don't want to expose SQL errors. This would also be a Very Good Place to put some logging to see if we do run into unexpected baddies.

Reply:
Yeah, I think there might be a problem with this: You're swallowing whatever specific SQLAlchemy error was raised and its specific error message and traceback replacing it with a generic `"The update was not successful"`. When trying to debug something in production (e.g. looking at Sentry logs) or in dev it won't be possible to see what actually happened.

We could log the original `SQLAlchemyError` before raising our own, as you say.

But I wonder if _not_ catching `SQLAlchemyError` here at all, and letting any unexpected `SQLAlchemyError` fly, is better? This way our approach to not letting `SQLAlchemyError`s leak out of the services code would be:

1. For specific SQLAlchemy errors that we _know_ might happen (e.g. the `IntegrityError` above) we catch them specifically and raise our own exception classes.
2. For unexpected SQLAlchemy errors that we haven't accounted for we do not catch them, let them fly and let them crash the request. The original SQLAlchemy error will be reported to Sentry and printed to dev consoles.

SQLAlchemy errors should never be part of a service function's public interface. Hence why we catch `IntegrityError` and raise our own `ConflictError`, and our docstring states that `ConflictError` might be raised. The unexpected SQLAlchemy errors that might be raised are _not_ part of our service's public interface -- they're bugs. They should _not_ be mentioned in the docstring. They should never be raised. If one does get raised it indicates a bug that needs to be fixed. So they should be uncaught.